kernel/cClosure.ml

```diff
@@ -368,7 +368,7 @@ and fterm =
   | FCoFix of cofixpoint * fconstr subs
   | FCaseT of case_info * constr * fconstr * constr array * fconstr subs (* predicate and branches are closures *)
   | FLambda of int * (Name.t * constr) list * constr * fconstr subs
-  | FProd of Name.t * fconstr * fconstr
+  | FProd of Name.t * constr * constr * fconstr subs
```

Comment:
I don't completely understand the code here, but why is it `constr * constr * fconstr subs` rather than `fconstr * constr * fconstr subs`? My understanding is that the first argument (after the name is the type), that isn't open, and we lose sharing (the purpose of `fconstr` if I understand correctly) if we convert it back into a `constr` during reduction.

All this said, this seems to be the way that lambda and case work, so...I'm not certain.